SIM: Implement simulation

jpanikulam opened the pull request:
See [Simulation wiki page](https://github.com/uf-mil/Sub8/wiki/Simulation) for considerations,

and [Simul8 wiki page](https://github.com/uf-mil/Sub8/wiki/Simul8) for what is in, and what is not.

The sim is now functional enough to test controllers and mappers, and therefore satisfies the first simulation design milestone. Refs #4 

Since this is already nearly 1,000 lines I'm submitting a PR now to receive early criticism.

Annhiluc committed:
SIM: Add physics for Simul8

jpanikulam committed:
SIM: Add shader parsing

jpanikulam committed:
SIM: Add sub-sim package structure

jpanikulam committed:
SIM: Add world rendering with diffuse lighting

jpanikulam committed:
SIM: Add visualization sandbox

jpanikulam committed:
SIM: Add Blinn-phong lighting

jpanikulam committed:
SIM: Add physics abstraction tools

jpanikulam committed:
SIM: Add add_plane to rendering.world

jpanikulam committed:
SIM: Implement functional simulation

jpanikulam committed:
SIM: Add simple sub8 physics entity

- Using box to simulate behavior
- Using Annie's adaptation of Forrest's thrust input

jpanikulam committed:
SIM: Clean up rendering

jpanikulam committed:
SIM: Add sub truth odometry publisher

jpanikulam committed:
SIM: Add buoyancy model to physics

- Also removed add_box and replaced it with add_entity

pemami4911 commented:
@zachgoins adding you to review this PR as well since it's a big one

pemami4911 commented:
Would be helpful to reviewers if a quick explanation of how this code was tested/verified was included in the commit message

pemami4911 commented on simulation/sub8_simulation/launch/sim_thrusters.launch:
```diff
@@ -0,0 +1,67 @@
+<launch>
```
This looks like the same thing we have over in the sub8_thruster_mapper. Is there a way to prevent duplicating code?

jpanikulam committed:
SIM: Address comments on PR #17

jpanikulam commented:
Addressing comments:
- Only some sim components will be configurable, unless we decide otherwise
- Cutting down on repeated code in the physics components is difficult, because ODE is picky about how geometries and bodies are defined
- I eliminated repeated code in the launch files

zachgoins commented:
I have looked through all the files and cannot see any glaring problems just by looking. It would be very beneficial though to include a few lines at the top of each file to give an idea of what the goal of the file is in the grand scheme of the sim.

zachgoins commented:
Is is possible to run instances of the sim without rendering graphics? We could use that feature to run continuous integration tests in the background with the physics alone.

jpanikulam commented:
That is a good point. 

Right now, physics is completely decoupled from rendering, and this is a good argument to keep it that way. 

@annhiluc thoughts?
- Jake

> On Sep 22, 2015, at 12:07 PM, zachgoins notifications@github.com wrote:
> 
> Is is possible to run instances of the sim without rendering graphics? We could use that feature to run continuous integration tests in the background with the physics alone.
> 
> —
> Reply to this email directly or view it on GitHub.

jpanikulam commented:
I suppose we'd also want to simulate vision in that case. We may have to fork Vispy to make it possible. 
- Jake

> On Sep 22, 2015, at 12:07 PM, zachgoins notifications@github.com wrote:
> 
> Is is possible to run instances of the sim without rendering graphics? We could use that feature to run continuous integration tests in the background with the physics alone.
> 
> —
> Reply to this email directly or view it on GitHub.

Annhiluc commented:
Since physics is independent of rendering, we could make an instance without visualization. What would need to be displayed, if anything?

fnivek commented:
Anyone know about the Model View Controller design pattern? This seems like
a perfect place to use the MVC pattern.

On Tue, Sep 22, 2015 at 12:53 PM, Annie Luc notifications@github.com
wrote:

> Since physics is independent of rendering, we could make an instance
> without visualization. What would need to be displayed, if anything?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/uf-mil/Sub8/pull/17#issuecomment-142347046.

## 

Sincerely,
Kevin French,
University of Florida,
4th year Electrical Engineer,
Cell: Available upon request,
Email: fnivek@gmail.com

jpanikulam commented:
Reassigning to Patrick for final pull
